### PR TITLE
[BE] Implement Project API endpoints

### PR DESCRIPTION
**Pull request by SummerTinio**
# App updates
- **Documentation** on Confluence: https://onramp-training.atlassian.net/l/c/QzoLtoTj
  - TODO: transfer to Swagger Docs
- **Implemented Project CRUD endpoints**
  - **Public** endpoints /api/projects/[id] and /api/projects
  - **Private** endpoints /api/users/projects/[id] and /api/users/projects routes **integrated with verifyRequest to require authorization (uid)**
  - **Fixed stalled responses** from server by sending JSON response on success & on fail
  - Implement **Error-propagation with Clearer Error boundaries** (ProjectService --> ProjectController --> Project API handler)
    - note: **This was necessary to identify which part of the API needed fixing**

Notes:
- Necessary steps for future troubleshooting:
  - Use the **most current .env file (with PUBLIC_API_URL_DEV)**
  - Set req.headers('Authorization') to Bearer `<AuthToken>` on Postman to make authorized-only requests (DELETE, PUT, POST)
    - Auth token expires after some time, so it may **need to be refreshed with a new auth token**
  - To **cascade deletion of relations upon deleting** table row (e.g. on Project table), **set { onDelete: 'CASCADE' }** on corresponding Relation (e.g. on Like & Investment entity)

### CODE REVIEW

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+import { NextApiRequest } from "next";
+
+export interface CreateParamsInterface {
+  title: string;
+  category: string;
+  description: string;
+  fundTiers: number[];
+  currFundGoal: number;
+  userId: number; 
+  // Is launchdate required soon as proj is being created?
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:14]: remove comment

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  // if not, pls remove the next line.
+  launchDate?: string;
+}
+
+export interface ProjectCreateApiRequest extends NextApiRequest {
+  // req.body.createParams.description
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:20]: remove commented code

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  body: CreateParamsInterface;
+}
+
+/**
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:24]: remove commented code

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+export interface FindByIdParamsInterface {
+  id: string;
+}
+
+export interface ProjectFindByIdApiRequest extends NextApiRequest {
+  body: FindByIdParamsInterface;
+}
+
+export interface UpdateByIdParamsInterface {
+  // id: string; <-- auto passed by query
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:46]: remove commented code

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  project: number;
+  pictures?: string[];
+  title?: string;
+  category?: string;
+  description?: string;
+  fundTiers?: number[];
+  currFundGoal?: number;
+  // comments?: Comment[]; <--- take this* out in other files
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:54]: remove commented code

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  id: string;
+}
+
+export interface ProjectDeleteByIdApiRequest extends NextApiRequest {
+  body: DeleteByIdParamsInterface;
+}
+
+export const SortByString = {
+  NEWEST: "createdDate",
+  TRENDING: "likesAmt", // TODO: Add column you want to sort by (instead of "likesAmt") @Pran
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:76]: add a Jira ticket here i.e. "TODO (pran) JIRA-1234: description"

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  RECOMMENDED: "user.id", // TODO: Add column you want to sort by (instead of "user.id") @Tapa
+};
+
+export enum SortByStringEnum {
+  NEWEST = "createdDate",
+  TRENDING = "likesAmt", // TODO: Add column you want to sort by (instead of "likesAmt") @Pran
+  RECOMMENDED = "user.id", // TODO: Add column you want to sort by (instead of "user.id") @Tapa
+}
+// SortByString.NEWEST
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:85]: remove commented code

File: src/backend/common/ProjectRequestApiInterfaces.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    | SortByStringEnum.NEWEST
+    | SortByStringEnum.TRENDING
+    | SortByStringEnum.RECOMMENDED;
+};
+
+export interface ProjectSortByApiRequest extends NextApiRequest {
+  body: SortByParamsType;
+}
+
+/**
```
Comment (sheikhrachel) [inline src/backend/common/ProjectRequestApiInterfaces.ts:100]: remove commented code

File: src/backend/common/RequestMethod.ts [hunk 9]
```diff
@@ -1,13 +1,13 @@
 const RequestMethod = {
```
Comment (sheikhrachel) [inline src/backend/common/RequestMethod.ts:1]: can't this be `export default RequestMethod = {`

File: src/backend/common/SortByConfig.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+const sortByString = {
```
Comment (sheikhrachel) [inline src/backend/common/SortByConfig.ts:1]: same here

File: src/backend/config/db.ts [hunk 11]
```diff
@@ -1,16 +1,16 @@
...
+import { Like } from "../entities/Like";
+import { Category } from "../entities/Category";
 
-import chalk from 'chalk';
+import chalk from "chalk";
 
 //create typeorm database connection using ormconfig.json file
 const connection = async () => {
-  console.log(chalk.green('Connecting to database...'));
+  console.log(chalk.green("Connecting to database..."));
```
Comment (sheikhrachel) [inline src/backend/config/db.ts:13]: remove console log

File: src/backend/config/db.ts [hunk 12]
```diff
@@ -21,12 +21,12 @@ const connection = async () => {
   const temp = await createConnection({
     type: process.env.DB_TYPE as any,
     url: process.env.DB_CONNECTION_STRING,
-    //logging: true, //statements/queries console logged to terminal -> should remove when not debugging
+    // logging: true, //statements/queries console logged to terminal -> should remove when not debugging
```
Comment (sheikhrachel) [inline src/backend/config/db.ts:24]: remove commented code

File: src/backend/config/db.ts [hunk 13]
```diff
@@ -21,12 +21,12 @@ const connection = async () => {
...
     synchronize: true, //will tralslate logic to sql
     entities: [User, Project, Investment, Comment, Like, Category],
   });
   if (getConnection().isConnected) {
-    console.log(chalk.green('DB Connected!'));
+    console.log(chalk.green("DB Connected!"));
```
Comment (sheikhrachel) [inline src/backend/config/db.ts:29]: remove console log

File: src/backend/controller/project/ProjectController.ts [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+      throw err;
+    }
+  },
+
+  // READ - all
+  findAll: async (req) => {
+    try {
+      return await ProjectService.findAll();
+    } catch (err) {
+      console.warn(
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:55]: same here

File: src/backend/controller/project/ProjectController.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    try {
+      if (Object.keys(req.query).length === 1) {
+        return await ProjectService.findById(req.query);
+      } else {
+        throw new BadRequestError(
+          "req.query is missing id (project id) - Did you forget to add the id on the endpoint? e.g. /api/project/10"
+        );
+      }
+    } catch (err) {
+      console.warn(
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:73]: and here

File: src/backend/controller/project/ProjectController.ts [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+            "req.query is missing id (project id) - Did you forget to add the id on the endpoint? e.g. /api/project/10"
+          );
+        }
+
+        if (Object.keys(req.body).length < 7) {
+          throw new BadRequestError("req.body has missing parameters.");
+        }
+      }
+    } catch (err) {
+      console.warn(
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:111]: and here

File: src/backend/controller/project/ProjectController.ts [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+      );
+      throw err;
+    }
+  },
+
+  // UPDATE - 1
+  updateById: async (req) => {
+    try {
+      if (
+        req.user === undefined ||
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:84]: repeated code.  can be encapsulated in a func that returns a boolean to check this logic

File: src/backend/controller/project/ProjectController.ts [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+        req.user.uid === undefined ||
+        req.user.uid === null
+      ) {
+        throw new AuthorizationError(
+          "Auth token does not exist or has expired"
+        );
+      }
+
+      if (
+        isAllTruthy(req.body) &&
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:94]: similar enough to L33 that you could create a shared validation method that returns a bool

File: src/backend/controller/project/ProjectController.ts [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+      );
+      throw err;
+    }
+  },
+
+  // DESTROY - 1
+  deleteById: async (req) => {
+    try {
+      if (
+        req.user === undefined ||
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:122]: same reusable method

File: src/backend/controller/project/ProjectController.ts [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+
+      if (Object.keys(req.query).length === 1) {
+        return await ProjectService.deleteById(parseInt(req.query.id));
+      } else {
+        throw new BadRequestError(
+          "req.query is missing id (project id) - Did you forget to add the id on the endpoint? e.g. /api/project/10"
+        );
+      }
+    } catch (err) {
+      console.warn(
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:143]: same console consideration

File: src/backend/controller/project/ProjectController.ts [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+        chalk.bgRed(`Error caught at ProjectController - ${err.message}`)
+      );
+      throw err;
+    }
+  },
+
+  // SORT - all (TODO: add logic to ProjectService.sortBy -- Tapa & Pran)
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:150]: same todo syntax as before

File: src/backend/controller/project/ProjectController.ts [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+  sortBy: async (req) => {
+    try {
+      if (isAllTruthy(req.body) && Object.keys(req.body).length === 1) {
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:153]: another potential to reuse the validation logic

File: src/backend/controller/project/ProjectController.ts [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+        return ProjectService.sortBy(req);
+      } else {
+        throw new Error(
+          "req.body is missing sortBy params! (i.e. SortByString.NEWEST, or SortByString.RECOMMENDED, or SortByString.TRENDING"
+        );
+      }
+    } catch (err) {
+      console.warn(
```
Comment (sheikhrachel) [inline src/backend/controller/project/ProjectController.ts:161]: another console consideration

File: src/backend/entities/User.ts [hunk 24]
```diff
@@ -1,39 +1,46 @@
...
 
-  @Column('text')
+  // this solved the 500
+  @Column({ type: "text", nullable: true })
   uid: string;
-
-  @Column('text')
+  /** 
+   @Column("text")
+   uid: string;
```
Comment (sheikhrachel) [inline src/backend/entities/User.ts:17]: remove commented code

File: src/backend/lib/prepareDbConnection.ts [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+  // if no connection exists yet
+  if (!connectionReadyPromise) {
+    // create one
+    connectionReadyPromise = (async () => {
+      try {
+        const staleConnection = getConnection();
+        // and close it (to clean up references to outdated db connections)
+        await staleConnection.close();
+      } catch (err) {
+        console.warn(err.message);
```
Comment (sheikhrachel) [inline src/backend/lib/prepareDbConnection.ts:21]: same console consideration

File: src/backend/middleware/verify_request.ts [hunk 26]
```diff
@@ -1,36 +1,38 @@
...
 }
 
 export const verifyRequest = (handler) => {
   return async (req: Request, res: NextApiResponse) => {
-    const accessToken = req.headers['authorization'];
-    let token: string = '';
+    const accessToken = req.headers["authorization"];
+    let token: string = "";
+
+    console.log(accessToken);
```
Comment (sheikhrachel) [inline src/backend/middleware/verify_request.ts:16]: remove this

File: src/backend/middleware/verify_request.ts [hunk 27]
```diff
@@ -1,36 +1,38 @@
...
 
     try {
       if (!accessToken) {
         console.log(
```
Comment (sheikhrachel) [inline src/backend/middleware/verify_request.ts:20]: remove this

File: src/backend/middleware/verify_request.ts [hunk 28]
```diff
@@ -1,36 +1,38 @@
...
         const decoded = await admin.auth().verifyIdToken(token);
         if (!decoded) {
-          throw new AuthorizationError('Invalid token');
+          throw new AuthorizationError("Invalid token");
         }
         req.user = decoded;
       }
     } catch (error) {
       if (error instanceof AuthorizationError) {
         console.log(
```
Comment (sheikhrachel) [inline src/backend/middleware/verify_request.ts:34]: remove this

File: src/backend/common/RequestMethod.ts (hunk 9 above)
Comment (SummerTinio) [reply src/backend/common/RequestMethod.ts:1]: tried this, but it needs an identifier before using default exports (unless using it as a named export)
image 
image

File: src/backend/common/SortByConfig.ts (hunk 10 above)
Comment (SummerTinio) [reply src/backend/common/SortByConfig.ts:1]: this also needs an identifier before using default exports (unless using it as a named export)
image
image

File: src/backend/controller/project/ProjectController.ts (hunk 22 above)
Comment (SummerTinio) [reply src/backend/controller/project/ProjectController.ts]: Extracted this logic to reusable utility function (commit dd8132e94796682b52d1583815d9ece9af7e14bb)

File: src/backend/controller/project/ProjectController.ts [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+        isAllTruthy(createParamsWithUid) &&
+        Object.keys(createParamsWithUid).length >= 7
+      ) {
+        const createdProj = await ProjectService.create(createParamsWithUid);
+        return [createdProj, StatusCodes.CREATED];
+      } else {
+        throw new BadRequestError("Missing params on req.body");
+      }
+    } catch (err) {
+      console.warn(
```
[the remark this answers is not in the document]
Comment (SummerTinio) [reply src/backend/controller/project/ProjectController.ts]: Removed all Project Controller console.warn statements on commit 3576cab033b8db764ec01516ca615e46ad90da6c

File: src/backend/entities/User.ts (hunk 24 above)
Comment (SummerTinio) [reply src/backend/entities/User.ts]: Removed commented code on User entity on commit 1c23c89080c0348a4fe2db7417719bc27500b96f

File: src/backend/lib/prepareDbConnection.ts (hunk 25 above)
Comment (SummerTinio) [reply src/backend/lib/prepareDbConnection.ts]: Removed console logs from other files- Deleted this file (prepareDbConnection.ts) per coordination with Pran, will be using `@backend/config/db.ts` instead. commit 6598e252e64c2c9560d11744911d7b1ed2b81428

File: src/backend/middleware/verify_request.ts (hunk 28 above)
Comment (SummerTinio) [reply src/backend/middleware/verify_request.ts:34]: Removed all console statements on verifyRequest Middleware- commit bf2df4e0913b967a58244ed8c0d4d41d8a922920

Comment (SummerTinio) [reply src/backend/middleware/verify_request.ts]: Removed all console statements on verifyRequest Middleware-  commit bf2df4e0913b967a58244ed8c0d4d41d8a922920

File: src/frontend/context/AuthProvider.tsx [hunk 30]
```diff
@@ -61,37 +61,47 @@ const reducer = (state, action) => {
...
         getUser();
-        getUserInvestments();
+        // getUserInvestments();
       }
     });
   }, []);
 
   const getUser = async () => {
-    const response = await axios.get('/users/get');
+    console.log(`we're in the getUser function!`);
```
Comment (sheikhrachel) [inline src/frontend/context/AuthProvider.tsx:81]: remove console.log

File: src/frontend/context/AuthProvider.tsx [hunk 31]
```diff
@@ -61,37 +61,47 @@ const reducer = (state, action) => {
...
+    const response = await axios.get("/users/get");
+    /**
```
Comment (sheikhrachel) [inline src/frontend/context/AuthProvider.tsx:83]: remove commented code

File: src/backend/services/db/ProjectService.ts [hunk 32]
```diff
@@ -0,0 +1,233 @@
...
+        throw new DatabaseError("Database connection failed");
+      }
+
+      const foundProject = await db
+        .createQueryBuilder()
+        .select("*")
+        .from("project", "project")
+        .where(`project.id = ${findByIdParams.id}`)
+        .getRawOne();
+      // .execute(); <-- doesnt work. only works with getRawOne
```
Comment (sheikhrachel) [inline src/backend/services/db/ProjectService.ts:132]: remove

File: src/frontend/context/AuthProvider.tsx [hunk 33]
```diff
@@ -61,37 +61,47 @@ const reducer = (state, action) => {
...
-        delete axios.defaults.headers.common['Authorization'];
+        router.push("/");
+        delete axios.defaults.headers.common["Authorization"];
       } else {
         const token = await getIdToken(user);
-        axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
+        axios.defaults.headers.common["Authorization"] = `Bearer ${token}`;
         getUser();
-        getUserInvestments();
+        // getUserInvestments();
```
Comment (sheikhrachel) [inline src/frontend/context/AuthProvider.tsx:75]: remove

File: src/backend/utils/existsInDb.ts [hunk 34]
```diff
@@ -0,0 +1,22 @@
...
+  id: number
+): Promise<boolean> {
+  const db = await getConnection();
+  const result = await db
+    .createQueryBuilder()
+    .select("*")
+    .from(entity, entityName)
+    .where(`${entityName}.id = ${id}`)
+    .getRawOne(); // .getOne() works
+  console.log(chalk.blueBright(result !== undefined));
```
Comment (sheikhrachel) [inline src/backend/utils/existsInDb.ts:17]: remove

File: src/backend/services/db/project/project_db.ts [hunk 35]
```diff
@@ -0,0 +1,108 @@
...
+      res.status(200).json({
+        userData,
+      });
+    } else {
+      res.status(404).json({
+        message: "Project not found",
+      });
+    }
+  } catch (error) {
+    console.log("ERROR: getProject() in project_db.ts", error);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:27]: throw the err don't log it

File: src/backend/services/db/project/project_db.ts [hunk 36]
```diff
@@ -0,0 +1,108 @@
...
+          email: email,
+          avatar: "",
+          bio: "",
+          investedAmt: 0,
+        },
+      ])
+      .execute();
+    return userData;
+  } catch (error) {
+    console.warn(error.message);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:51]: remove console.warn

File: src/backend/services/db/project/project_db.ts [hunk 37]
```diff
@@ -0,0 +1,108 @@
...
+      res.status(200).json({
+        message: "Project updated",
+      });
+    } else {
+      res.status(404).json({
+        message: "Project not updated",
+      });
+    }
+  } catch (error) {
+    console.log(error);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:81]: throw the err don't log it

File: src/backend/services/db/project/project_db.ts [hunk 38]
```diff
@@ -0,0 +1,108 @@
...
+      .update(Project)
+      .set({
+        first_name: req.body.firstName,
+        last_name: req.body.lastName,
+        avatar: req.body.avatar,
+        bio: req.body.bio,
+      })
+      .where("email = :email", { email })
+      .execute();
+    console.log(userData);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:70]: remove

File: src/backend/services/db/project/project_db.ts [hunk 39]
```diff
@@ -0,0 +1,108 @@
...
+  const db = await connection();
+  try {
+    const email = req.user.email;
+    const userData = await db
+      .createQueryBuilder()
+      .delete()
+      .from(Project)
+      .where("email = :email", { email })
+      .execute();
+    console.log(userData);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:95]: remove

File: src/backend/services/db/project/project_db.ts [hunk 40]
```diff
@@ -0,0 +1,108 @@
...
+      res.status(200).json({
+        message: "Project deleted",
+      });
+    } else {
+      res.status(404).json({
+        message: "Project not deleted",
+      });
+    }
+  } catch (error) {
+    console.log(error);
```
Comment (sheikhrachel) [inline src/backend/services/db/project/project_db.ts:106]: throw the err don't log it

File: src/backend/services/db/ProjectService.ts [hunk 41]
```diff
@@ -0,0 +1,233 @@
...
+      const db = await connection();
+
+      if (!existInDb("project", Project, projectId)) {
+        throw new NotFoundError("Project does not exist");
+      }
+      const deletedProject = await db
+        .createQueryBuilder()
+        .delete()
+        .from(Project)
+        .where("id = :id", { id: projectId }) //<--
```
Comment (sheikhrachel) [inline src/backend/services/db/ProjectService.ts:191]: remove comment

File: src/backend/middleware/verify_request.ts (hunk 28 above)
Comment (sheikhrachel) [reply src/backend/middleware/verify_request.ts]: just throw the error, there's no need to log it for a user to see

File: src/backend/services/db/project/project_db.ts (hunk 37 above)
Comment (SummerTinio) [reply src/backend/services/db/project/project_db.ts]: replaced this file with ProjectService.ts and removed console statements on commit 680934a6ed27920720e4b650ab52462dbcae02da

File: src/backend/utils/existsInDb.ts (hunk 34 above)
Comment (SummerTinio) [reply src/backend/utils/existsInDb.ts]: Removed commented code & console log on commit 20ab2de035d6ce4190a597d456e85349cc5a5967

File: src/backend/services/db/ProjectService.ts (hunk 32 above)
Comment (SummerTinio) [reply src/backend/services/db/ProjectService.ts]: Removed commented code on commit 0805dae1223f68e8bac9b6b6560bec4a066bbd45

File: src/frontend/context/AuthProvider.tsx (hunk 31 above)
Comment (SummerTinio) [reply src/frontend/context/AuthProvider.tsx]: Removed all commented code and logs on AuthProvider on commit d5fa007f58982dc041beb6d67032e19bdb0df66a

File: src/backend/services/db/ProjectService.ts (hunk 41 above)
Comment (SummerTinio) [reply src/backend/services/db/ProjectService.ts]: Removed comment on commit 7b3114900797f1387ddc0f6ac92839b4bdda3598

**Pull request merged**